feat: Add `simultaneousWithExternalGesture` to ReanimatedSwipable

MrSltun opened the pull request:
## Description
This PR adds a new prop to `RenimatedSwipable` that allows adding a gesture config that can be used simultaneously with `ReanimatedSwipeable`'s gesture config.

The new prop is called `simultaneousWithExternalGesture`, it uses the the cross-component interaction methodology that's mentioned in the [documentation here](https://docs.swmansion.com/react-native-gesture-handler/docs/fundamentals/gesture-composition/#simultaneouswithexternalgesture).

I also update the docs for it

### Motivation
I've been recently using ReanimatedSwipeable component, and I tried to add an additional gesture config to enable haptic feedback and "release to run function" kind of interaction

After looking through the issues in the repo, I found this issue #2862, I tried to apply it, but it turned out that it was only applicable to the Swipable component

I tried to find a way to add a gesture config that would work simultaneously with the swipeable one but I couldn't find a way to add it. So I looked through the documentation of RNGH for simultaneous gesture handlers and came up with this solution

<!--
Description and motivation for this PR.

Include 'Fixes #<number>' if this is fixing some issue.
-->

## Test plan
I already tested it locally on both iOS and Android, and the callbacks work as expected

Not sure if there's something else to do here since this is my first contribution

<!--
Describe how did you test this change here.
-->

latekvo commented:
I think this is very reasonable.

As far as I know, it's currently not possible to write interactions with components that utilise `Gestures` within themselves, without passing some references to the component. I think this is something worth looking into. 

One thing i'd request, is that you rename the prop to `simultaneousWithExternalGesture`, as we already use `simultaneousWithExternalGesture` for this functionally identical purpose.

I'm not sure if we want to add `requireExternalGestureToFail` and `blocksExternalGesture`, since user usually has access to these function on both the parent, and the child, while I think adding them to `Swipeable` would only allow the user to use them on the child (`Swipeable`) while not allowing for them to be used by the parent (`Gesture` that is wrapping `Swipeable`). Let's wait for @m-bert's and @j-piasecki's opinions on this.

<details>
<summary>
Collapsed Pan + Tap test
</summary>

```tsx
import React from 'react';
import { StyleSheet, Text, View } from 'react-native';
import { Gesture, GestureDetector } from 'react-native-gesture-handler';

export default function EmptyExample() {
  const pan = Gesture.Pan().onStart(() => console.log('Pan'));
  const tap = Gesture.Tap().onStart(() => console.log('Tap'));
  return (
    <GestureDetector gesture={pan}>
      <GestureDetector gesture={tap}>
        <View style={styles.container}>
          <Text style={{ fontSize: 64, opacity: 0.25 }}>🎯</Text>
        </View>
      </GestureDetector>
    </GestureDetector>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
});

```

</details>


<details>
<summary>
Collapsed Pan + Tap + Swipeable test
</summary>

```tsx
import React from 'react';
import { StyleSheet, Text } from 'react-native';
import { Gesture, GestureDetector } from 'react-native-gesture-handler';
import ReanimatedSwipeable from 'react-native-gesture-handler/ReanimatedSwipeable';
import Animated, {
  SharedValue,
  useAnimatedStyle,
} from 'react-native-reanimated';

function LeftAction(_prog: SharedValue<number>, drag: SharedValue<number>) {
  const styleAnimation = useAnimatedStyle(() => {
    return {
      transform: [{ translateX: drag.value - 50 }],
    };
  });

  return (
    <Animated.View style={styleAnimation}>
      <Text style={styles.leftAction}>Text</Text>
    </Animated.View>
  );
}

function RightAction(_prog: SharedValue<number>, drag: SharedValue<number>) {
  const styleAnimation = useAnimatedStyle(() => {
    return {
      transform: [{ translateX: drag.value + 50 }],
    };
  });

  return (
    <Animated.View style={styleAnimation}>
      <Text style={styles.rightAction}>Text</Text>
    </Animated.View>
  );
}

export default function EmptyExample() {
  const pan = Gesture.Pan().onStart(() => console.log('Pan'));
  const tap = Gesture.Tap().onStart(() => console.log('Tap'));
  return (
    <GestureDetector gesture={pan}>
      <GestureDetector gesture={tap}>
        <ReanimatedSwipeable
          containerStyle={styles.swipeable}
          friction={2}
          leftThreshold={80}
          enableTrackpadTwoFingerGesture
          rightThreshold={40}
          renderLeftActions={LeftAction}
          renderRightActions={RightAction}>
          <Text>[Reanimated] Swipe me!</Text>
        </ReanimatedSwipeable>
      </GestureDetector>
    </GestureDetector>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  swipeable: {
    height: 50,
    backgroundColor: 'papayawhip',
    alignItems: 'center',
  },
  leftAction: {
    width: 50,
    height: 50,
    backgroundColor: 'crimson',
  },
  rightAction: {
    width: 50,
    height: 50,
    backgroundColor: 'purple',
  },
});

```

</details>

m-bert commented:
Hi @MrSltun! Thanks for this PR!

> One thing i'd request, is that you rename the prop to simultaneousWithExternalGesture, as we already use simultaneousWithExternalGesture for this functionally identical purpose, [more info here](https://docs.swmansion.com/react-native-gesture-handler/docs/fundamentals/gesture-composition/#simultaneouswithexternalgesture).

Actually I'm not sure if we want to rename it in current state. Also, now when I look at it, I assume that `ReanimatedSwipeable` is wrapped with `GestureDetector`, which means that `simultaneousGesture` is used inside another `GestureDetector`. For that cases we have `simultaneousWithExternalGesture` prop, as combining gestures in this way may not work. 

What I think should be done here is to call this prop `simultaneousWithExternalGesture` and use it inside:

```diff
    const panGesture = useMemo(
      () =>
        Gesture.Pan()
          .enabled(enabled !== false)
          .enableTrackpadTwoFingerGesture(enableTrackpadTwoFingerGesture)
          .activeOffsetX([-dragOffsetFromRightEdge, dragOffsetFromLeftEdge])
          .onStart(() => {
            updateRightElementWidth();
          })
          .onUpdate(
            (event: GestureUpdateEvent<PanGestureHandlerEventPayload>) => {
              userDrag.value = event.translationX;

              const direction =
                rowState.value === -1
                  ? SwipeDirection.RIGHT
                  : rowState.value === 1
                    ? SwipeDirection.LEFT
                    : event.translationX > 0
                      ? SwipeDirection.RIGHT
                      : SwipeDirection.LEFT;

              if (!dragStarted.value) {
                dragStarted.value = true;
                if (rowState.value === 0 && onSwipeableOpenStartDrag) {
                  runOnJS(onSwipeableOpenStartDrag)(direction);
                } else if (onSwipeableCloseStartDrag) {
                  runOnJS(onSwipeableCloseStartDrag)(direction);
                }
              }

              updateAnimatedEvent();
            }
          )
          .onEnd(
            (event: GestureStateChangeEvent<PanGestureHandlerEventPayload>) => {
              handleRelease(event);
            }
          )
          .onFinalize(() => {
            dragStarted.value = false;
          })
+       .simultaneousWithExternalGesture(simultaneousGesture),
      [
        dragOffsetFromLeftEdge,
        dragOffsetFromRightEdge,
        dragStarted,
        enableTrackpadTwoFingerGesture,
        enabled,
        handleRelease,
        onSwipeableCloseStartDrag,
        onSwipeableOpenStartDrag,
        rowState,
        updateAnimatedEvent,
        updateRightElementWidth,
        userDrag,
      ]
    );
```

Could you please check if this works?

MrSltun commented:
>Actually I'm not sure if we want to rename it in current state. Also, now when I look at it, I assume that ReanimatedSwipeable is wrapped with GestureDetector, which means that simultaneousGesture is used inside another GestureDetector. For that cases we have simultaneousWithExternalGesture prop, as combining gestures in this way may not work.

My Approach was just to combine an external gesture with the `ReanimatedSwipeable` gesture (without a parent `GestureDetector` wrapping the `ReanimatedSwipeable`)

But I like the approach of having the `ReanimatedSwipeable ` gesture working simultaneously with a parent `GestureDetector`, I will update the code, test it, and report back here :)

latekvo commented:
Works great, test code used: 
<details>
<summary>
Collapsed repro
</summary>

```tsx
import React from 'react';
import { View, Text } from 'react-native';
import { Gesture, GestureDetector } from 'react-native-gesture-handler';
import Swipeable from 'react-native-gesture-handler/ReanimatedSwipeable';

export default function Example() {
  const pan = Gesture.Pan()
    .onBegin(() => console.log('began'))
    .onStart(() => console.log('active'))
    .onFinalize((_, success) => console.log(success ? 'ended' : 'cancelled'));

  return (
    <GestureDetector gesture={pan}>
      <Swipeable
        leftThreshold={50}
        rightThreshold={50}
        renderLeftActions={() => {
          return (
            <View style={{ height: 80, width: 80, backgroundColor: 'yellow' }}>
              <Text>Left</Text>
            </View>
          );
        }}
        renderRightActions={() => {
          return (
            <View style={{ height: 80, width: 80, backgroundColor: 'magenta' }}>
              <Text>Right</Text>
            </View>
          );
        }}>
        <View
          style={{
            height: 80,
            backgroundColor: 'blue',
          }}
        />
      </Swipeable>
    </GestureDetector>
  );
}

```

</details>

Just one thing - the `Tap` in `Swipeable` isn't made simultaneous with the provided `simultaneousWithExternalGesture` prop. Unless I'm missing something, I think it should also be made simultaneous.

Also, I'm getting a type error when i pass my `pan` gesture to the `simultaneousWithExternalGesture` in the test code I provided, but as far as I see, `GestureType` is the correct type to use in this scenario. I'm not sure if this is a case of package version discrepancy or some other issue?
![image](https://github.com/user-attachments/assets/5277bbca-6c01-4e42-950c-8c4838d0d418)

MrSltun commented:
Thank you for testing the changes @latekvo 🫶 

> Just one thing - the `Tap` in `Swipeable` isn't made simultaneous with the provided `simultaneousWithExternalGesture` prop. Unless I'm missing something, I think it should also be made simultaneous.

Hmm I tested the pan gestures, but didn't check the other types, I'll test all the gestures and see which works and which doesn't, and try to figure out why


> Also, I'm getting a type error when i pass my pan gesture to the simultaneousWithExternalGesture in the test code I provided, but as far as I see, GestureType is the correct type to use in this scenario. I'm not sure if this is a case of package version discrepancy or some other issue?

Oh that's weird! I didn't come across it, not sure why it's happening but I'll check

m-bert commented:
## TypeScript error

We've spent some time with @tjzel trying to find out what's going on. Turns out that this is a problem with our configuration (thank you @j-piasecki ❤️). 

Could you please update `tsconfig.json` in our example app? Otherwise CIs may fail

```diff
{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "noEmit": true,
    "baseUrl": ".",
    "paths": {
      "react-native-gesture-handler": ["../src/index.ts"],
+     "react-native-gesture-handler/ReanimatedSwipeable": [
+       "../src/components/ReanimatedSwipeable.tsx"
+     ],
      "react-native-gesture-handler/jest-utils": ["../src/jestUtils/index.ts"]
    }
  },
  "include": ["src/**/*.ts", "src/**/*.tsx", "App.tsx"]
}

```

## Types

Regarding the code, I'd like to use the type that `simultaneousWithExternalGesture` expects, i.e. 

```tsx
simultaneousWithExternalGesture: Exclude<GestureRef, number>[];
```

And that requires a change in passing it into config:

```tsx
.simultaneousWithExternalGesture(
  ...(simultaneousWithExternalGesture ?? [])
),
```

In that case, example usage looks like this:

```tsx
<ReanimatedSwipeable
  simultaneousWithExternalGesture={[pan, tap]}
  ...
>
```

I know that this may be inconvenient with single gestures - we can change the type to:

```tsx
simultaneousWithExternalGesture: Exclude<GestureRef, number> | Exclude<GestureRef, number>[];
```

This will require a bit more logic when passing this prop into config, but it will be easier to use 😅

MrSltun commented:
@m-bert thank you for taking the time to test the change :)

I updated the the example app's tsconfig

Re: updating the `simultaneousWithExternalGesture` type,
I like the new approach since it allows multiple gestures to be passed to the `ReanimatedSwipeable` gesture handler :)

I updated the name as well to reflect that change to be `simultaneousWithExternalGestures`

I also tested the new change and it works as it should 🎉

m-bert commented:
Hi @MrSltun! Thanks for updating `tsconfig`. Could you please share code that you've used to test these changes? You can either add it into _**Test plan**_ section, or here in comments.

m-bert commented on src/components/ReanimatedSwipeable.tsx:
```diff
@@ -202,6 +203,12 @@ export interface SwipeableProps
+   * An array of gesture objects containing the configuration and callbacks to be
+   * used with the swipeable's gesture handler.
+   */
+  simultaneousWithExternalGestures?: Exclude<GestureRef, number>[];
```
```suggestion
  simultaneousWithExternalGesture?: Exclude<GestureRef, number>[];
```

m-bert reviewed (changes requested):
Hi @MrSltun! Thank you for changing our `tsconfig` 😅 

### Prop name

Please change prop name to `simultaneousWithExternalGesture`. I know that it sounds better with `s` at the end, but this is how it is currently named in Gesture Handler and we would like to stick to it.

### Prop type

After looking at it I'd prefer to remove array if not necessary. It means that type of the prop should be

```tsx
simultaneousWithExternalGesture?:
  | Exclude<GestureRef, number>
  | Exclude<GestureRef, number>[];
```

To handle it correctly, I suggest adding the following logic into `useMemo`:

```js
      ...
        .onFinalize(() => {
          dragStarted.value = false;
        });

      if (!simultaneousWithExternalGesture) {
        return pan;
      }

      if (Array.isArray(simultaneousWithExternalGesture)) {
        pan.simultaneousWithExternalGesture(...simultaneousWithExternalGesture);
      } else {
        pan.simultaneousWithExternalGesture(simultaneousWithExternalGesture);
      }

      return pan;
```

### Tap gesture

Currently this prop **_is not_** used with `Tap` gesture that is present inside `Swipeable` - please add this 😅

MrSltun commented:
@m-bert Thank you for the detailed feedback :) 

I completely missed the tab gesture 😅 thank you for pointing it out!

I just updated the name of the variable, as well as the definition of the pan gesture, and I added the gesture to the tab one as well :)

m-bert reviewed (approved):
Thank you very much for this PR @MrSltun ❤️
